applications/HDF5Application/custom_utilities/data_type_utilities.h

```diff
@@ -88,7 +90,89 @@ void GetShapeFromAttributes(
     }
 }
 
+template<class TContainerType>
+[[nodiscard]] std::vector<std::string> GetListOfAvailableVariables(
+    TContainerType& rContainer,
+    const DataCommunicator& rDataCommunicator)
+{
+    KRATOS_TRY
+
+    using map_type = std::unordered_map<VariableData::KeyType, std::string>;
+
+    // Map to Map reduction class
+    class MapReduction
+    {
+    public:
+        using return_type = map_type;
+
+        return_type mValue;
+
+        /// access to reduced value
+        return_type GetValue() const
+        {
+            return mValue;
+        }
+
+        /// NON-THREADSAFE (fast) value of reduction, to be used within a single thread
+        void LocalReduce(map_type rValue){
+            mValue.merge(rValue);
+        }
+
+        /// THREADSAFE (needs some sort of lock guard) reduction, to be used to sync threads
+        void ThreadSafeReduce(MapReduction& rOther)
+        {
+            KRATOS_CRITICAL_SECTION
+            mValue.merge(rOther.mValue);
+        }
+    };
 
+    // collect all the unique key, variable names map from all the entities
+    const auto& key_variable_names_map = block_for_each<MapReduction>(rContainer, map_type(), [](auto& rEntity, auto& rTLS) {
```

Comment:
I know you have to return the map because of how our reduction system works, but then you should make absolutely sure that you're not accidentally copying it every time.
```suggestion
    const auto& key_variable_names_map = block_for_each<MapReduction>(rContainer, map_type(), [](const auto& rEntity, auto& rTLS) -> map_type& {
```